### PR TITLE
Start implementing browser tests using karma

### PR DESCRIPTION
**Pull request by Dignifiedquire**
Ref #77 

- [x] Travis happy
  - [x] `test:node`
  - [x] `test:browser`
- [x] Use Karma :)
- [x] refactor tests to pass in nodejs environment
- [x] add saucelabs (or similar) integration (account created, need to configure it now)

### CODE REVIEW

**Comment by Dignifiedquire** [comment]
@VictorBjelkholm nearly done with all your listed todos ;)

**Comment by Dignifiedquire** [comment]
At the moment the `test:browser` script fails on travis because it can not run `node_modules/.bin/ipfs daemon` for some reason. If anyone has an idea please help, I'm at a loss here

**Comment by jbenet** [comment]
i dont see the logging output from `ipfs` but i would suspect a port clash

**Comment by Dignifiedquire** [comment]
All done @diasdavid @VictorBjelkholm please give this a code review :)

File: karma.conf.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
+module.exports = function (config) {
+  if (!process.env.SAUCE_USERNAME || !process.env.SAUCE_ACCESS_KEY) {
```
Comment (VictorBjelkholm) [inline karma.conf.js]: I'll need this to be able to run in local? I think this should be optional

File: karma.conf.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    console.log('Make sure the SAUCE_USERNAME and SAUCE_ACCESS_KEY environment variables are set.')
+    process.exit(1)
+  }
+
+  // Browsers to run on Sauce Labs
+  // Check out https://saucelabs.com/platforms for all browser/OS combos
+  var customLaunchers = {
```
Comment (VictorBjelkholm) [inline karma.conf.js]: Should probably have something running on Linux as well, just in case. Might be overkill though

File: package.json [hunk 3]
```diff
@@ -25,15 +25,24 @@
...
+    "karma-mocha": "^0.2.0",
+    "karma-sauce-launcher": "^0.3.0",
+    "mocha": "^2.3.3",
     "pre-commit": "^1.0.6",
     "standard": "^5.2.2",
     "uglify-js": "^2.4.24"
   },
   "scripts": {
-    "test": "./node_modules/.bin/mocha",
+    "test": "ls -la node_modules/.bin && npm run test:node && npm run test:browser",
```
Comment (VictorBjelkholm) [inline package.json]: This ts command should be here I guess, no?

Comment (Dignifiedquire) [inline package.json]: yeah that was just for debugging :)

File: karma.conf.js (hunk 1 above)
Comment (Dignifiedquire) [inline karma.conf.js]: Right, not sure of the best way to do that, suggestions?

File: karma.conf.js (hunk 2 above)
Comment (Dignifiedquire) [inline karma.conf.js]: Launcher list should be bigger anyway probably, but just wanted to add sth for now

File: karma.conf.js (hunk 1 above)
Comment (VictorBjelkholm) [inline karma.conf.js]: Extract the SauceLabs specific parts that only get added if `process.env.SAUCE_USERNAME` and `SAUCE_ACCESS_KEY` is set would be a simple way to achieve it.

File: karma.conf.js (hunk 2 above)
Comment (VictorBjelkholm) [inline karma.conf.js]: Fair enough

**Comment by diasdavid** [comment]
this is awesome! Thank you @Dignifiedquire and @VictorBjelkholm ! :D

I'm getting 2 tests fail though, one might be related with being in a slow wifi, but the other (when daemon stops) shouldn't throw. Am I the only one seeing this? 

```bash
» npm test                                                                                                                                       ◉ ◼◼◼◼◼◼◼◼◼◼

> ipfs-api@2.4.1 test /Users/david/Documents/code/ipfs/node-ipfs-api
> npm run test:node && npm run test:browser


> ipfs-api@2.4.1 test:node /Users/david/Documents/code/ipfs/node-ipfs-api
> mocha



  ipfs node api
    # ipfs node setup
    # ipfs init done
    # ipfs daemon running
    ✓ has the api object
    ✓ add file
    ✓ add buffer
    - add path
    ✓ cat
    1) ls
    ✓ config set/get
    ✓ block.put
    ✓ block.get
    ✓ object.put
    ✓ object.get
    ✓ object.data
    - refs
    ✓ id
    - returns an error when getting a non-existent key from the DHT
    - puts and gets a key value pair in the DHT
    2) test for error after daemon stops


  11 passing (16s)
  4 pending
  2 failing

  1) ipfs node api ls:
     Error: timeout of 10000ms exceeded. Ensure the done() callback is being called in this test.


  2) ipfs node api test for error after daemon stops:
     Error: the object {
  "Code": 0
  "Message": "context canceled"
  "uncaught": true
} was thrown, throw an Error :)





npm ERR! Darwin 15.0.0
npm ERR! argv "/usr/local/bin/iojs" "/usr/local/bin/npm" "run" "test:node"
npm ERR! node v4.2.0
npm ERR! npm  v3.3.6
npm ERR! code ELIFECYCLE
npm ERR! ipfs-api@2.4.1 test:node: `mocha`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the ipfs-api@2.4.1 test:node script 'mocha'.
npm ERR! This is most likely a problem with the ipfs-api package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     mocha
npm ERR! You can get their info via:
npm ERR!     npm owner ls ipfs-api
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/david/Documents/code/ipfs/node-ipfs-api/npm-debug.log
npm ERR! Test failed.  See above for more details.
```

**Comment by Dignifiedquire** [comment]
@diasdavid run them a couple of times, no idea they sometimes work and sometimes fail for me I think it's not dependent on this api but rather on ipfs answering fast enough. Maybe the daemon checks should be separated out or so.

**Comment by diasdavid** [comment]
Question posted on IRC: 

```
14:09 <+daviddias> dignifiedquire: I'm not being able to understand why the 'add path' test stays on 'pending' state
14:09 <+daviddias> seems that tests are being executed in parallel now 
14:09 <+daviddias> and since others fail, that one doesn't even finish
```


Also, running npm run test:browser throws because it is not using a disposable Node:

```
» npm run test:browser                                                                                                                          ◉ ◼◼◼◼◼◼◼◼◼◻

> ipfs-api@2.4.1 test:browser /Users/david/Documents/code/ipfs/node-ipfs-api
> ipfs init && API_ORIGIN="*" ./node_modules/.bin/concurrent --kill-others "./node_modules/.bin/ipfs daemon" "./node_modules/.bin/karma start --single-run=true karma.conf.js"

initializing ipfs node at /Users/david/.ipfs
Error: ipfs configuration file already exists!
Reinitializing would overwrite your keys.
(use -f to force overwrite)


npm ERR! Darwin 15.0.0
npm ERR! argv "/usr/local/bin/iojs" "/usr/local/bin/npm" "run" "test:browser"
npm ERR! node v4.2.0
npm ERR! npm  v3.3.9
npm ERR! code ELIFECYCLE
npm ERR! ipfs-api@2.4.1 test:browser: `ipfs init && API_ORIGIN="*" ./node_modules/.bin/concurrent --kill-others "./node_modules/.bin/ipfs daemon" "./node_modules/.bin/karma start --single-run=true karma.conf.js"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the ipfs-api@2.4.1 test:browser script 'ipfs init && API_ORIGIN="*" ./node_modules/.bin/concurrent --kill-others "./node_modules/.bin/ipfs daemon" "./node_modules/.bin/karma start --single-run=true karma.conf.js"'.
npm ERR! This is most likely a problem with the ipfs-api package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     ipfs init && API_ORIGIN="*" ./node_modules/.bin/concurrent --kill-others "./node_modules/.bin/ipfs daemon" "./node_modules/.bin/karma start --single-run=true karma.conf.js"
npm ERR! You can get their info via:
npm ERR!     npm owner ls ipfs-api
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/david/Documents/code/ipfs/node-ipfs-api/npm-debug.log
```

**Comment by diasdavid** [comment]
@Dignifiedquire @VictorBjelkholm would you mind if I add a branch to node-ipfs-api that we all can work in (giving you perm) to do the PR to master, so that we don't have to create forks to PR into @Dignifiedquire fork so that it gets PR'ed into master?

**Comment by Dignifiedquire** [comment]
sure thing

**Comment by diasdavid** [comment]
Awesome! :) Check https://github.com/ipfs/node-ipfs-api/pull/81

closing this one :)

**Pull request closed without merging**